File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == nullptr){
            return list2;
        } else if(list2 == nullptr){
            return list1;
        }
        ListNode* currNode1 = list1;
        ListNode* currNode2 = list2;
        ListNode* newHead = nullptr ;
        if(currNode1->val <= currNode2->val){
            newHead = currNode1;
            currNode1 = currNode1->next;
        }else{
            newHead = currNode2;
            currNode2 = currNode2->next;
        }
        ListNode* currNewNode = newHead;
        while(currNode1 != nullptr && currNode2 != nullptr){
            if(currNode1->val <= currNode2->val){
            currNewNode->next = currNode1;
            currNode1 = currNode1->next;
        }else{
            currNewNode->next = currNode2;
            currNode2 = currNode2->next;
        }
            currNewNode = currNewNode->next;
        }

        while(currNode1 != nullptr){
            currNewNode->next = currNode1;
            currNode1 = currNode1->next;
            currNewNode = currNewNode->next;
        }
        while(currNode2 != nullptr){
            currNewNode->next = currNode2;
            currNode2 = currNode2->next;
            currNewNode = currNewNode->next;
        }

        return newHead;
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not l1:
            return l2
        elif not l2:
            return l1
        resultNode = ListNode()
        start = resultNode
        carry = 0
        while l1 or l2 or carry != 0:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0
            sum = num1 + num2 + carry
            carry = sum // 10
            val = sum % 10
            newNode = ListNode(val)
            resultNode.next = newNode
            resultNode = resultNode.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return start.next


